FV variable support for ThermochimicaNodalData

parikshitbajpai opened the pull request:
Closes #25661
Requires https://github.com/ORNL-CEES/thermochimica/pull/149

parikshitbajpai committed:
Use `setDofValue` instead of `setNodalValue` in `ThermochimicaNodalData`
(#25661)

parikshitbajpai committed:
`MooseWritableVariable` type for `Coupleable::writableVariable` (#25661)

`MooseWritableVariable` is of the type `MooseVariableField<Real>` and extends the applicability to `writableVariable` which would return a reference to `MooseVariable` type. This change allows `writableVariable` to be extended for use with FV variables

parikshitbajpai committed:
Add FV type to `ChemicalCompositionAction` (#25661)

parikshitbajpai committed:
Tests for `writableVariable` on FV variables (#25661)

- Switches to direct call to `Coupleable::getVarHelper<MooseWritableVariable>` in `Coupleable::writableVariable`
- Added `test/tests/auxkernels/nodal_aux_var/multi_update_fv_test.i`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1808612/) on eb99229 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>e8c7f0</th>
      <th colspan=4>#25663 eb9922</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.69%</td>
      <td>85.69%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>91797</td>
      <td>91797</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15335</td>
      <td>15335</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25663/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25663/coverage/framework)


## Modules coverage

### Chemical reactions

<table>
  <thead>
    <tr>
      <th></th>
      <th>e8c7f0</th>
      <th colspan=4>#25663 eb9922</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>92.20%</td>
      <td>92.21%</td>
      <td>+0.01%</td>
      <td>80.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1301</td>
      <td>1302</td>
      <td>+1</td>
      <td>12</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>110</td>
      <td>110</td>
      <td>-</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25663/coverage_diff/chemical_reactions)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25663/coverage/chemical_reactions)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25663/coverage/xfem)

</details>

## Warnings

- `chemical_reactions` new line coverage rate 80.00% is less than the suggested 90.0%



This comment will be updated on new commits.

dschwen commented on modules/chemical_reactions/include/actions/ChemicalCompositionAction.h:
```diff
@@ -9,6 +9,7 @@
 
 #pragma once
 
+#include "AddVariableAction.h"
```
Just move this to the .C file

dschwen reviewed (commented):
Looks fine to me, but coverage is an issue. This might be tough to address, as we are intercepting Thermochimica errors that we do not know how to trigger in the first place.

lindsayad reviewed (changes requested):
This looks good to me as well

GiudGiud commented on test/tests/auxkernels/nodal_aux_var/tests:
```diff
@@ -29,8 +29,7 @@
     issues = '#2099'
-    requirement = "The AuxKernel objects shall be capable of wriyting to to multiple coupled "
-                  "variables."
+    requirement = "The AuxKernel objects shall be capable of writing to multiple coupled variables."
```
```suggestion
    requirement = "Auxiliary kernel objects shall be capable of writing to multiple coupled variables."
```

parikshitbajpai committed:
Fix typos in test requirements to address review comments (#25661)

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>
Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

parikshitbajpai committed:
Additional testing to capture Thermochimica failure cases (#25661)

parikshitbajpai commented:
> Looks fine to me, but coverage is an issue. This might be tough to address, as we are intercepting Thermochimica errors that we do not know how to trigger in the first place.

I figured out a way to trigger those errors and that should increase coverage. Waiting on https://github.com/ORNL-CEES/thermochimica/pull/149 before I push as I'll need to update Thermochimica submodule.

GiudGiud commented:
when you squash the two commits please add something to the message
some classics : fix typos / adapt tests requirement / add more constness

we try to keep the commit history very informative

parikshitbajpai committed:
Update Thermochimica submodule (#25661)

This submodule update fixes a couple of incorrect conditionals in Thermochimica Leveling solver which cause malloc errors down the line

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1815386/) on 4fa4228 : invalidated by @lindsayad
